### ISSUE TITLE
Panic trying to read settings.toml from 64 bit NFS mount on 32 bit linux

### ISSUE THREAD

**Issue by Tyg13** [rust-lang/rustup#2254]
**Problem**
Can't install or use `rustup` when on `Linux rhel5 2.6.18-8.el5 i686`. When running any command that loads `settings.toml` (like `rustup update`), a run-time panic occurs.

<details>
  <summary>Backtrace</summary>

``` error: could not read settings file: '/home/LOCAL/tla/.local/.rustup/settings.toml'
error: caused by: thread 'main' panicked at 'strerror_r failure', src/libstd/sys/unix/os.rs:111:13
stack backtrace:
   0:   0x917c00 - backtrace::backtrace::libunwind::trace::h1f4a6004caa6702e
                       at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1:   0x917c00 - backtrace::backtrace::trace_unsynchronized::hc3a0416ecea81263
                       at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2:   0x917c00 - std::sys_common::backtrace::_print_fmt::h6172996b7af07508
                       at src/libstd/sys_common/backtrace.rs:77
   3:   0x917c00 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h6b904f28bad2cada
                       at src/libstd/sys_common/backtrace.rs:61
   4:   0x5cf12c - core::fmt::write::h8650968343dd62c9
                       at src/libcore/fmt/mod.rs:1028
   5:   0x9172dc - std::io::Write::write_fmt::hb64f100f52f7ea6a
                       at src/libstd/io/mod.rs:1412
   6:   0x916ed6 - std::sys_common::backtrace::_print::hb968ae4db43a5949
                       at src/libstd/sys_common/backtrace.rs:65
   7:   0x916ed6 - std::sys_common::backtrace::print::h352e70faf9696ad9
                       at src/libstd/sys_common/backtrace.rs:50
   8:   0x916ed6 - std::panicking::default_hook::{{closure}}::h8fb02f3abc190d44
                       at src/libstd/panicking.rs:188
   9:   0x916588 - std::panicking::default_hook::ha2267c3d3607d470
                       at src/libstd/panicking.rs:205
  10:   0x916588 - std::panicking::rust_panic_with_hook::h627ca090d036cfc9
                       at src/libstd/panicking.rs:464
  11:   0x915c00 - std::panicking::begin_panic::hd857b4812a1b286c
                       at src/libstd/panicking.rs:400
  12:   0x91aaef - std::sys::unix::os::error_string::h1d382ef36116680c
                       at src/libstd/sys/unix/os.rs:111
  13:   0x91a809 - <std::io::error::Error as core::fmt::Display>::fmt::h886342b9be50b9ec
                       at src/libstd/io/error.rs:533
  14:   0x4d07c3 - <&T as core::fmt::Display>::fmt::hf5c783236aa5a299
  15:   0x5cf12c - core::fmt::write::h8650968343dd62c9
                       at src/libcore/fmt/mod.rs:1028
  16:   0x4e1c91 - rustup_init::log::err_fmt::h6c16c12556e70eb3
  17:   0x4ead7b - rustup_init::common::report_error::hfce30ee56beb2fdf
  18:   0x54501e - rustup_init::main::hfa2f5d877ac46169
  19:   0x4cf806 - std::rt::lang_start::{{closure}}::hdb5faaf5c16a8cd8
  20:   0x92cf76 - std::rt::lang_start_internal::{{closure}}::{{closure}}::h2f63d790ead9148c
                       at src/libstd/rt.rs:48
  21:   0x92cf76 - std::sys_common::backtrace::__rust_begin_short_backtrace::h2a41dc8aeff9d971
                       at src/libstd/sys_common/backtrace.rs:129
  22:   0x5508c2 - main
  23:   0x171dec - __libc_start_main
```

</details>

**Steps**

1. `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | RUST_BACKTRACE=full sh -s -- -y --default-toolchain stable`

**Possible Solution(s)**
My understanding is that this is a 64 bit inode support issue. The path I am attempting to use as my `RUSTUP_HOME`  is on an NFS mount with 64 bit inodes. If I set `RUSTUP_HOME` to my local filesystem, then the install finishes with no issue. This only occurs on 32 bit linux, where it is necessary to set `-D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1` when interacting with filesystems with 64 bit inodes.

**Notes**

Output of `rustup --version`:
`rustup 1.21.1 (7832b2ebe 2019-12-20)`
Output of `rustup show`:
```
error: could not read settings file: '/home/LOCAL/tla/.local/.rustup/settings.toml'
error: caused by: thread 'main' panicked at 'strerror_r failure', src/libstd/sys/unix/os.rs:111:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

**Comment by Tyg13** [comment]
This is probably a much larger issue, perhaps with `libstd` or `libc` upstream. Perhaps it would be smarter to file this bug there.

**Comment by rbtcollins** [comment]
I agree that this is upstream of rustup; I think it going to be a libstd bug ultimately.

**Comment by kinnison** [comment]
Given that we state that the Linux target is meant to support `2.6.18+` this is likely to indeed by upstream of us.  As you've noted, `libc` might have issues, and that isn't bound by the rust project's platform support quite so strictly, but I'd be very pleased if that issue gets some attention

**Comment by Tyg13** [comment]
This was actually a bug with my kernel patch version, fixed in later versions of RHEL 5.2. As such, the correct fix is to patch my kernel. This doesn't appear to be an issue with Rust itself.
